fix: Exotel Call Log fixes

nemesis189 opened the pull request:
- When creating a call the status was always set as ringing even for missed calls, fixed to show the correct status:
   **No Answer** if no agent picks up the call
   **Canceled** if the customer hangs up before any agent picks up the call 

- Added field to display name of employee to whom the call is directed towards

nemesis189 committed:
fix: call status fix

nemesis189 committed:
fix: added employee name to call log

nemesis189 committed:
fix: added field to show called group, user_id

nemesis189 committed:
fix: added type of call select field, additional status for agent rejecting call

nemesis189 committed:
fix: sider fixes

surajshetty3416 commented on erpnext/erpnext_integrations/exotel_integration.py:
```diff
@@ -33,10 +33,25 @@ def handle_end_call(**kwargs):
 def handle_missed_call(**kwargs):
-	update_call_log(kwargs, 'Missed')
+	status = ""
+	CallType = kwargs.get("CallType")
```
```suggestion
	call_type = kwargs.get("CallType")
```
according to our naming convention, we use CapitalCase for class names or for query builder instances and snake case for variables

surajshetty3416 commented on erpnext/public/js/call_popup/call_popup.js:
```diff
@@ -140,6 +140,15 @@ class CallPopup {
+				'fieldname': 'type_of_call',
+				'label': 'Type Of Call',
+				'fieldtype': 'Select',
+				'options': '\nFrappe Cloud Queries/Plan\nEnterprise Plans\nPartnership\nSupport\nBilling/Renewal\nOpen source / Junk',
```
👀

"Frappe Cloud Queries" or other options here are not really relevant for all telephony users... better to have a link field instead.

surajshetty3416 commented on erpnext/public/js/call_popup/call_popup.js:
```diff
@@ -140,6 +140,15 @@ class CallPopup {
 				'hide_border': 1,
+			}, {
+				'fieldname': 'type_of_call',
+				'label': 'Type Of Call',
```
```suggestion
				'label': 'Call Type',
```
simpler?

surajshetty3416 commented on erpnext/telephony/doctype/call_log/call_log.json:
```diff
@@ -134,15 +138,34 @@
+   "depends_on": "to",
+   "fieldname": "employee_call_directed_to",
+   "fieldtype": "Data",
+   "label": "Employee Call Directed To",
```
```suggestion
   "label": "Call Received By",
```
?

nemesis189 replied:
call may not always be 'received'(picked up/ attended) by the agent so i felt 'directed to' would make more sense, anyway I'll change

surajshetty3416 commented on erpnext/telephony/doctype/call_log/call_log.py:
```diff
@@ -33,6 +33,16 @@ def before_insert(self):
+			emp_number_reversed = (self.get("to"))[-1:-11:-1]
+			emp_number = emp_number_reversed[-1::-1]
+
+			employee = frappe.get_all("Employee", filters={"cell_number":["like","%"+emp_number+"%"]}, fields=["first_name", "middle_name", "last_name", "user_id"])
```
```suggestion
			employee = frappe.get_all("Employee", filters={
				"cell_number": ["like", "%"+emp_number+"%"]
			}, fields=["first_name", "middle_name", "last_name", "user_id"])
```

nemesis189 committed:
fix: call type doctype, fixes

nemesis189 committed:
Merge branch 'exotel-fixes' of https://github.com/nemesis189/erpnext into exotel-fixes

nemesis189 committed:
fix: linter, sider fixes

surajshetty3416 commented on erpnext/telephony/doctype/call_log/call_log.py:
```diff
@@ -33,6 +33,18 @@ def before_insert(self):
+		if self.is_incoming_call():
+			# Taking the last 10 digits of the number
+			emp_number_reversed = (self.get("to"))[-1:-11:-1]
+			emp_number = emp_number_reversed[-1::-1]
```
use `strip_number` instead? Also, we avoid short-forms for variable names... let's keep it `employee_number`.

nemesis189 replied:
Done

surajshetty3416 commented on erpnext/telephony/doctype/call_log/call_log.py:
```diff
@@ -33,6 +33,18 @@ def before_insert(self):
+			emp_number = emp_number_reversed[-1::-1]
+			employee = frappe.get_all("Employee", filters={
+				"cell_number": ["like", "%"+emp_number+"%"]
+			}, fields=["first_name", "middle_name", "last_name", "user_id"])
```
you can extend `get_employees_with_number` method to make this work.

nemesis189 replied:
tweaked and used the method as suggested

surajshetty3416 commented on erpnext/telephony/doctype/call_log/call_log.py:
```diff
@@ -33,6 +33,18 @@ def before_insert(self):
+				"cell_number": ["like", "%"+emp_number+"%"]
+			}, fields=["first_name", "middle_name", "last_name", "user_id"])
+
+			self.call_received_by = get_employee_name(employee[0])
```
Let's keep `call_rececived_by` a link field. We can also, enable the [title for link](https://frappeframework.com/docs/v13/user/en/basics/doctypes/form_&_view_settings#title-field) for Employee doctype.

surajshetty3416 commented:
@nemesis189 let's also introduce basic tests for this integration.

nemesis189 committed:
fix: used get_employees_with_number, strip_number methods

ankush committed:
Merge branch 'develop' into exotel-fixes

nemesis189 committed:
fix: added tests(fixed)

surajshetty3416 committed:
Merge branch 'develop' into exotel-fixes

surajshetty3416 committed:
test: Refactor exotel test setup

- Remove unnecessary code
- Move test data to separate file
- Make proper test assertions

surajshetty3416 committed:
chore: Remove unused code

- and simplify get_call_log

surajshetty3416 committed:
fix: Handle exception where no employee is returned

surajshetty3416 committed:
test: Fix erroneous code

surajshetty3416 committed:
test: Use FrappeAPITestCase to track coverage

surajshetty3416 committed:
Merge branch 'develop' into exotel-fixes

surajshetty3416 committed:
test: Fix used frappe._dict to avoid AttributeError down the line

ref: https://github.com/frappe/erpnext/runs/5816574721?check_suite_focus=true#step:12:880

surajshetty3416 committed:
Merge branch 'develop' into exotel-fixes

surajshetty3416 committed:
Merge branch 'develop' into exotel-fixes

surajshetty3416 committed:
test: Clean up form_dict

To avoid failures like
https://github.com/frappe/erpnext/runs/5887687369?check_suite_focus=true#step:12:783

surajshetty3416 committed:
Merge branch 'develop' into exotel-fixes

surajshetty3416 committed:
test: Do not overwrite frappe.form_dict to retain proxy reference

Set frappe.local.form_dict instead

surajshetty3416 committed:
fix: Update received_by if "to" is changed

surajshetty3416 commented:
@Mergifyio backport version-13-hotfix
